specgen/specgen/hellocap/listener.go

```diff
@@ -0,0 +1,200 @@
+package hellocap
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"net"
+	"net/http"
+	"sync"
+)
+
+const tlsRecordHeaderLen = 5
+
+type contextKey int
+
+// contextHelloKey is the context key for retrieving the ClientHello.
+const contextHelloKey contextKey = 0
+
+// FromRequest returns the TLS ClientHello sent by the remote. This request must have come through a
+// handler attached to the Server type in this package.
+func FromRequest(req *http.Request) ([]byte, error) {
+	hello, ok := req.Context().Value(contextHelloKey).(capturedHello)
+	if !ok {
+		return nil, errors.New("no attached ClientHello")
+	}
+	return hello.hello, hello.err
+}
+
+// attachHello is used to put a captured ClientHello on a http.Request context. This is how the
+// ClientHello, captured at the TCP level, is made available to users of this library.
+func attachHello(req *http.Request, hello capturedHello) *http.Request {
```

Comment:
This is a bit tortured. I'm not sure we care about matching the hellos to the request - at least, the calling code definitely doesn't. Do you have a use case for this in mind for the future? Absent any reason to match hellos with requests, I would do one of the following:

 - Just give `hellocap.Server` a `Hello([]byte)` hook at the top level, and pass that down into the connection
 - Give `hellocap.Server` an `io.Writer` or `os.File` to dump all the client hellos out to, separated by `\0` or something
 - Accumulate client hellos in a `[]byte[5]` or whatever, which is only accessible after the server is stopped

Reply:
This can be useful if one wants to collect, say, Aser-Agent (or other headers) used in this request, so it can be matched to ClientHello.
Sorry for interjection.

Reply:
We chatted about this on a call, but for posterity:

We do care about matching hellos to the request - this is how the `specgen` utility is able to respond to each request with the captured ClientHello.  Personally, I find this more user-friendly than pulling the hellos out of the logs (and correlating each based on timing).  One can also imagine running a web server which remote clients can hit to see their own captured ClientHello.

> This can be useful if one wants to collect, say, User-Agent (or other headers) used in this request, so it can be matched to ClientHello.

This is a good point!  Thank you @fedosgad